### PR TITLE
Refactor Os::FileSystem and Os::Directory

### LINKED ISSUE #303: Svc::FileManager MoveFile command does not work across partitions
This command invokes the native OS rename interface. This doesn't work on a file system when the source and destination are on different file system partitions. It does work to a different directory on the same file system. Is there a different underlying call that more mimics the system "mv" command?

### LINKED ISSUE #2687: OSAL Refactor: Os::FileSystem and Os::Directory
| | |
|:---|:---|
|**_F´ Version_**| |
|**_Affected Component_**|  |
---
## Feature Description

Refactor the above OSAL pieces.  Some quick thoughts:

1. Directory should be merged into FileSystem
2. For namespace-bucked functions, do the following:
  1. Delegate  structure as normal
  2. static methods of class can delegate to singleton implementation.

### PR DESCRIPTION
**Pull request by thomas-bc**
| | |
|:---|:---|
|**_Related Issue(s)_**| https://github.com/nasa/fprime/issues/2687 |
|**_Has Unit Tests (y/n)_**| y |
|**_Documentation Included (y/n)_**| n |

---
## Change Description

Refactoring Os::FileSystem and Os::Directory to the new OSAL pattern

## Rationale

See https://github.com/nasa/fprime/issues/2687

## Future Work

Document the entire OSAL

### CODE REVIEW

File: Os/Directory.hpp [hunk 1]
```diff
@@ -1,44 +1,127 @@
-#ifndef _Directory_hpp_
-#define _Directory_hpp_
+#ifndef _OS_DIRECTORY_HPP_
+#define _OS_DIRECTORY_HPP_
```
Comment (Meganbrentshandacrain) [inline Os/Directory.hpp:2]: Megan crain

File: Os/Posix/Directory.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+#include <Fw/Types/Assert.hpp>
+#include <Os/Posix/Directory.hpp>
+#include <Os/Posix/error.hpp>
+#include <Fw/Types/StringUtils.hpp>
+
+namespace Os {
+namespace Posix {
+namespace Directory {
+
+PosixDirectory::PosixDirectory() : Os::DirectoryInterface(), m_handle() {}
```
[the remark this answers is not in the document]
Comment (Meganbrentshandacrain) [reply Os/Posix/Directory.cpp:18]: > ## More than one statement per line
> 
> This line contains 2 statements; only one is allowed.
> 
> [Show more details](https://github.com/nasa/fprime/security/code-scanning/75291)

- [ ] Os/Posix/Directory.cpp

**Review by Meganbrentshandacrain — APPROVED** [review]

File: Os/CMakeLists.txt [hunk 3]
```diff
@@ -97,26 +98,33 @@ endif()
...
   "${CMAKE_CURRENT_LIST_DIR}/test/ut/OsSystemResourcesTest.cpp"
   "${CMAKE_CURRENT_LIST_DIR}/test/ut/OsMutexBasicLockableTest.cpp"
 )
 register_fprime_ut()
 if (BUILD_TESTING)
-    foreach (TEST IN ITEMS StubFileTest PosixFileTest StubMutexTest PosixMutexTest) # add? : StubTaskTest PosixTaskTest
-       if (TARGET "${TEST}")
-           add_dependencies("Os_ut_exe" "${TEST}")
-       endif()
+    foreach (TEST IN ITEMS
```
Comment (LeStarch) [inline Os/CMakeLists.txt:119]: We need to take an action to clean this up.  Outside this PR.

File: Os/Posix/DefaultDirectory.cpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// ======================================================================
```
Comment (LeStarch) [inline Os/Posix/DefaultDirectory.cpp:1]: Directory, FileSystem, and File should all be in one default file and one implementation, no?

File: Svc/FileManager/FileManager.cpp [hunk 5]
```diff
@@ -62,8 +62,9 @@ namespace Svc {
   {
     Fw::LogStringArg logStringDirName(dirName.toChar());
     this->log_ACTIVITY_HI_CreateDirectoryStarted(logStringDirName);
+    bool errorIfDirExists = true;
     const Os::FileSystem::Status status =
-      Os::FileSystem::createDirectory(dirName.toChar());
+      Os::FileSystem::createDirectory(dirName.toChar(), errorIfDirExists);
```
Comment (LeStarch) [inline Svc/FileManager/FileManager.cpp:67]: Was the original implementation to error on exists?

File: config/FpConfig.h [hunk 6]
```diff
@@ -8,10 +8,11 @@
  * ALL RIGHTS RESERVED.  United States Government Sponsorship
  * acknowledged.
  */
-#include <Fw/Types/BasicTypes.h>
 #ifndef FPCONFIG_H_
 #define FPCONFIG_H_
 
+#include <Fw/Types/BasicTypes.h>
```
Comment (LeStarch) [inline config/FpConfig.h:14]: Why did this move?

File: Os/Posix/DefaultDirectory.cpp (hunk 4 above)
Comment (thomas-bc) [reply Os/Posix/DefaultDirectory.cpp:1]: https://github.com/nasa/fprime/issues/2905

File: Os/CMakeLists.txt (hunk 3 above)
Comment (thomas-bc) [reply Os/CMakeLists.txt:119]: https://github.com/nasa/fprime/issues/2904

File: Svc/FileManager/FileManager.cpp (hunk 5 above)
Comment (thomas-bc) [reply Svc/FileManager/FileManager.cpp:67]: Yes! Not having it made a couple UTs fail and that's what prompted me adding it to the interface as an optional parameter

**Comment by thomas-bc** [comment]
@LeStarch comments addressed and follow-up issues created. 
I'm ready to merge this when you are!

**Review by LeStarch — APPROVED** [review]

**Pull request merged**